
# Introduction

## Segmentation

## Yeast Data

chemostat oscillations: transcriptome oscillates, stable over various 
conditions, also seen in batch, and single cells from un-synchronized
chemostat and in microfluidics

here, RNA-seq over three cycles of the short period IFO strain, thus 
at unprecedented temporal resolution; unexpected singularity
during third cycle

# Methods
## Problem & Algorithm: `Karl`

## Implementation: `segmenTier`

* `segmenTier`
** Algorithm: R & Rcpp
** Parameter scanning framework
** Plot functions 

* `segmenTools`, library & command-line scripts
** presegment.R: 
   pre-segmentation by distance cutoffs
** runSegmentier.R: 
   command-line wrapper; used for parallelization on SGE cluster
** coor2index maps to yeast R64-1-1 genome; taken from `genomeBrowser`
** parameter scan analysis framework:
   criteria: jaccard, query/target length vs. test set (ORF transcripts); 
   segment length distribution, segment phase and amplitude


## Pre-segmentation

## Parameter Scan

* Segment length distributions
* ORF Recovery: coding vs. non-coding, red vs. blue genes
** Time-series pre-processing: ash trafo of DC 
** Exponent `E` and nuissance cluster correlation `nui`

* Examples: 
** SRG1/SER3, unknown at GDH3!
* domain_046_69kb: good sense:antisense osci at cluster A gene ECM32 (DNA dependent ATPase/DNA helicase!), surrounded by D and B oscillators
** SPO7/FUN4 (primsegv3 00016) & APD1/SPP381 (primsegv3 00100)
   -> difficult separation of close in-phase transcripts,

## Segment Dynamics

* coding vs. non-coding: 
