---
title: "segmenTier: Similarity-Based Segmentation of Multi-Dimensional Signals"
author: "Rainer Machne, Douglas B. Murray, Peter F. Stadler"
date: "`r Sys.Date()`"
output: 
    rmarkdown::html_vignette:
      toc: true
bibliography: segmenTier.bib
vignette: >
  %\VignetteIndexEntry{segmenTier}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

![](logo.png) 

`segmenTier` is a dynamic programming solution to segmentation based
on maximization of arbitrary similarity measures within segments
as developed in @Machne2017.

In addition to the core algorithm, the package provides time-series
processing and clustering functions as described in the
publication. These are generally applicable where a `k-means`
clustering yields meaningful results, and have been specifically
developed for clustering of the Discrete Fourier Transform of periodic
gene expression data ("circadian" or "yeast metabolic oscillations").

This clustering approach is outlined in the supplemental material of
@Machne2012, and here is used as a basis of
segment similarity measures.  Note, that the functions
`processTimeseries` and `clusterTimeseries`, can also be used as
stand-alone tools for periodic time-series analysis and clustering.


# Theory

`segmenTier`'s input is a clustering
$\mathcal{C}_{\alpha}\subseteq\mathbb{X}$, $\alpha=1,\dots N$.
`segmenTier` then solves the recursion:

\begin{equation}
  S_{k,\alpha} =
  \max_{j\le k} \max_{\beta\ne\alpha}
                 S_{j-1,\beta} + s(j,k,\alpha) - M\;,
\label{eq:01}
\end{equation}

where $s(j,k,\alpha)$ is a scoring function that provides similarity of a segment
from position $j$ to $k$ to clusters $\mathcal{C}_\alpha$, *e.g.* based on Pearson correlation
to cluster centers, and $M$ is a penalty incurred
by the use of a new segment, a fixed cost for each jump that allows to 
fine-tune minimal segment lengths. Backtracing the maximal scores $S(k,\alpha)$ then
provides both segment borders and segment cluster association.

Clustering tools are provided in the package, and a full pipeline is
available for clustering the Fourier-transform of oscillatory RNA-seq time-series,
("circadian or metabolic oscillations"). See **Quick Guide**.
 

The theory behind the package is detailed in @Machne2017. 

# Usage

## Installation

From CRAN:

```{r, eval=FALSE}
install.packages("segmenTier")
```

The development version can be obtained from github using [`devtools`](https://cran.r-project.org/package=devtools):

```{r, eval=FALSE}
library(devtools)
install_github("raim/segmenTier", subdir = "pkg")
```



## Quick Guide

```{r, fig.width=7, fig.height=3}
library(segmenTier)

data(primseg436) # RNA-seq time-series data

# Fourier-transform and cluster time-series:
tset <- processTimeseries(ts=tsd, dft.range=1:7, dc.trafo="ash")
cset <- clusterTimeseries(tset, K=12)
# ... segment it:
segments <- segmentClusters(seq=cset, M=100, E=2, nui=3, S="icor")
# and inspect results:
plotSegmentation(tset,cset,segments)
print(segments)
## and get segment border table for further processing
head(segments$segments)
```



Usage of the package is further demonstrated in two R demos.

## Demo I: Direct Interface to Algorithm

The main low level interface to the algorithm, function
`segmentClusters`, is demonstrated in the file `demo/segment_test.R`.
It produces Supplemental Figure S1 of @Machne2017.

To run it as a demo in R simply type:

```{r, eval=FALSE}
demo("segment_test", package = "segmenTier")
```

## Demo II: Clustering, Batch Segmentation & Parameter Scans

A real-life data set is processed, clustered and segmented with
varying parameters in `demo/segment_data.R`.

This demo runs quite long, since it calculates many segmentations. It
provides a comprehensive overview of the effects of segmentation
parameters `E`, `M` and `nui`, and produces (among others) Figure 3
and Supplemental Figures S4a and S4b of @Machne2017.

```{r, eval=FALSE}
demo("segment_data", package = "segmenTier")
```

# References

